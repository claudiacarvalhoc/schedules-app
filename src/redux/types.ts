export const FETCH_SCHEDULE_SUCCESS = 'FETCH_SCHEDULE_SUCCESS';
export const FETCH_SCHEDULE_FAILURE = 'FETCH_SCHEDULE_FAILURE';

export const FETCH_SCHEDULE_LOGS_SUCCESS = 'FETCH_SCHEDULE_LOGS_SUCCESS';
export const FETCH_SCHEDULE_LOGS_FAILURE = 'FETCH_SCHEDULE_LOGS_FAILURE';

export const REQUEST_RETIRE_SCHEDULE = 'REQUEST_RETIRE_SCHEDULE';
export const REQUEST_UNRETIRE_SCHEDULE = 'REQUEST_UNRETIRE_SCHEDULE';
export const UPDATE_SELECTED_SCHEDULE = 'UPDATE_SELECTED_SCHEDULE';