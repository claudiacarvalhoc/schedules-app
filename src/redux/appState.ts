import { DefaultRootState } from "react-redux";

/**
 * The App Status
 */
export interface AppState extends DefaultRootState {
    status: AppStatusSectionsState;
    selectedScheduleId?: number;
    schedules: ScheduleState[];
    scheduleLogs: ScheduleLogState[];
    texts: {
        header: HeaderTextStatus;
        schedules: ScheduleTextState;
    }
}

/**
 * The status to fetch data per section
 */
export interface AppStatusSectionsState {
    schedules: StatusState;
    scheduleLogs: StatusState;
}

/**
 * The Boot Status is defined for options: Loading / Success / Failure.
 */
export enum StatusState {
    Loading,
    Success,
    Failure
}

/**
 * The Schedule interface.
 */
export interface ScheduleState {
    id: number;
    avatar: AvatarState;
    name: string;
    description: string;
    isRetired: boolean;
    tasksCount: number;
    startPoint: Date;
    endPoint: Date;
    dayOfWeek: number;
    dayOfMonth: number;
    startDate: Date;
    endDate: Date;
    schedules?: IntervalTypeState;
    timePeriod?: number;
}

/**
 * The Avatar Identification
 */
export interface AvatarState {
    letter: string;
    color: string;
}

/**
 * The Schedule Interval Types
 */
export enum IntervalTypeState {
    Once,
    Year,
    Week,
    Hour,
    Minute,
    Second
}

/**
 * The Schedule Logs interface.
 */
export interface ScheduleLogState {
    id: number;
    startTime: Date;
    endTime: Date;
    status: LogStatusState;
    serverName: string;
    scheduleId: number;
}

/**
 * The Schedule Logs Status
 */
export enum LogStatusState {
    Terminated,
    Pending,
    Completed,
    Exception,
    Running
}

/**
 * Texts used on Header
 * Note: This allow us to add localisation feature easily
 */
export interface HeaderTextStatus {
    titleText: string;
}

/**
 * Texts used on Schedule
 * Note: This allow us to add localisation feature easily
 */
export interface ScheduleTextState {
    buttonRetireText: string;
    buttonUnretireText: string;
    errorMessageText: string;
    buttonRetryText: string;
    emptyMessageText: string;
}
