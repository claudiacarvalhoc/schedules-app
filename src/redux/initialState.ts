import { AppState, Status } from "./appState";

/**
 * This is the initial state of redux.
 */
export const appInitialState: AppState = {
    status: {
        schedules: Status.Loading,
        scheduleLogs: Status.Loading,
    },
    selectedScheduleId: undefined,
    schedules: [],
    scheduleLogs: [],
};
