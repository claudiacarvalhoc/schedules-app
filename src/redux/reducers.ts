import { Reducer, combineReducers } from 'redux';
import { AppState } from './appState';
import { appReducer } from './app/reducer';

export const rootReducer = combineReducers({
    app: appReducer as Reducer<AppState>,
});

export type RootState = ReturnType<typeof rootReducer>;
export type DefaultRootState = ReturnType<typeof rootReducer>;