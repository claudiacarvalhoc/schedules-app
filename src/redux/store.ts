import { AnyAction } from 'redux';
import { AppState } from './appState';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { applyMiddleware, compose, createStore } from 'redux';
import { createLogger } from 'redux-logger';
import { rootReducer } from './reducers';
import thunk from 'redux-thunk';

export type AppThunk<R> = ThunkAction<R, AppState, unknown, AnyAction>;
export type AppDispatch = ThunkDispatch<AppState, unknown, AnyAction>;


export const configureStore = (initialState: AppState) => {
    const logger = createLogger();
    const composeEnhancers = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
    const middleware = [thunk, logger];

    const store = createStore(
        rootReducer,
        {
            app: initialState,
        },
        composeEnhancers(applyMiddleware(...middleware)),
    );

    return store;
};

export default configureStore;
