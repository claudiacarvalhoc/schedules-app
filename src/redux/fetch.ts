import { AppDispatch } from './store';

export const getSchedules = (): ((dispatch: AppDispatch) => Promise<void>) => {
    return dispatch => {
        // TODO : ....
        return Promise.resolve();
    };
};

export const getScheduleLogs = (): ((dispatch: AppDispatch) => Promise<void>) => {
    return dispatch => {
        // TODO : ....
        return Promise.resolve();
    };
};
