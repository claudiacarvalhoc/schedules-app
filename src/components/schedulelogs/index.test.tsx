import Schedules from './index';
import Enzyme, { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { appInitialState } from '../../redux/initialState';
import { LogStatusState, ScheduleLogState, StatusState } from '../../redux/appState';
import thunk from 'redux-thunk';

Enzyme.configure({ adapter: new Adapter() });

const scheduleLogItem: ScheduleLogState = {
    id: 1193520,
    startTime: new Date("2021-12-18T06:35:09.604Z"),
    endTime: new Date("2021-07-12T23:43:39.395Z"),
    status: LogStatusState.Running,
    serverName: "esse ad deserunt",
    scheduleId: 23802893,
};

const getMountWrapper = (scheduleLogsStatus: StatusState, scheduleLogsData: ScheduleLogState[] = []) => {
    const store = configureMockStore([thunk])({
      app: {
        ...appInitialState,
        scheduleLogs: scheduleLogsData,
        status: {
            ...appInitialState.status,
            scheduleLogs: scheduleLogsStatus
        },
      },
    });
    return mount(
        <Provider store={store}>
            <Schedules />
        </Provider>
    );
  };

describe('<ScheduleLogs />', () => {
    it('render correctly when component is loading', () => {
        const component = getMountWrapper(StatusState.Loading);
        expect(toJson(component)).toMatchSnapshot();
    });
    it('render correctly when schedule logs has an array data and data was saved sucessfully', () => {
        const component = getMountWrapper(StatusState.Success, [scheduleLogItem]);
        expect(toJson(component)).toMatchSnapshot();
    });
    it('render correctly when schedule logs is empty and data was saved sucessfully', () => {
        const component = getMountWrapper(StatusState.Success, []);
        expect(toJson(component)).toMatchSnapshot();
    });
    it('render correctly when schedule logs is undefined and data was saved sucessfully', () => {
        const component = getMountWrapper(StatusState.Success, []);
        expect(toJson(component)).toMatchSnapshot();
    });
    it('render correctly when schedule logs couldn\'t be fetched and data was saved as failure', () => {
        const component = getMountWrapper(StatusState.Failure);
        expect(toJson(component)).toMatchSnapshot();
    });
});
