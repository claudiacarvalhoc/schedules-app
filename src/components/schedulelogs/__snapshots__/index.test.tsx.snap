// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScheduleLogs /> render correctly when schedule logs couldn't be fetched and data was saved as failure 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ScheduleLogs)>
    <ScheduleLogs
      fetchScheduleLogs={[Function]}
      isSelected={false}
      logs={false}
      resetSelectedSchedule={[Function]}
      scheduleName={false}
      status={2}
      texts={
        Object {
          "buttonRetryText": "Retry",
          "emptyMessageText": "No schedules logs available.",
          "errorMessageText": "Something went wrong ...",
          "noSelectedScheduleText": "Please select a schedule.",
        }
      }
    >
      <div>
        <div
          className="failure"
        >
          <p>
            Something went wrong ...
          </p>
          <WithStyles(ForwardRef(Button))
            color="primary"
            disableElevation={true}
            onClick={[Function]}
            variant="contained"
          >
            <ForwardRef(Button)
              classes={
                Object {
                  "colorInherit": "MuiButton-colorInherit",
                  "contained": "MuiButton-contained",
                  "containedPrimary": "MuiButton-containedPrimary",
                  "containedSecondary": "MuiButton-containedSecondary",
                  "containedSizeLarge": "MuiButton-containedSizeLarge",
                  "containedSizeSmall": "MuiButton-containedSizeSmall",
                  "disableElevation": "MuiButton-disableElevation",
                  "disabled": "Mui-disabled",
                  "endIcon": "MuiButton-endIcon",
                  "focusVisible": "Mui-focusVisible",
                  "fullWidth": "MuiButton-fullWidth",
                  "iconSizeLarge": "MuiButton-iconSizeLarge",
                  "iconSizeMedium": "MuiButton-iconSizeMedium",
                  "iconSizeSmall": "MuiButton-iconSizeSmall",
                  "label": "MuiButton-label",
                  "outlined": "MuiButton-outlined",
                  "outlinedPrimary": "MuiButton-outlinedPrimary",
                  "outlinedSecondary": "MuiButton-outlinedSecondary",
                  "outlinedSizeLarge": "MuiButton-outlinedSizeLarge",
                  "outlinedSizeSmall": "MuiButton-outlinedSizeSmall",
                  "root": "MuiButton-root",
                  "sizeLarge": "MuiButton-sizeLarge",
                  "sizeSmall": "MuiButton-sizeSmall",
                  "startIcon": "MuiButton-startIcon",
                  "text": "MuiButton-text",
                  "textPrimary": "MuiButton-textPrimary",
                  "textSecondary": "MuiButton-textSecondary",
                  "textSizeLarge": "MuiButton-textSizeLarge",
                  "textSizeSmall": "MuiButton-textSizeSmall",
                }
              }
              color="primary"
              disableElevation={true}
              onClick={[Function]}
              variant="contained"
            >
              <WithStyles(ForwardRef(ButtonBase))
                className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-disableElevation"
                component="button"
                disabled={false}
                focusRipple={true}
                focusVisibleClassName="Mui-focusVisible"
                onClick={[Function]}
                type="button"
              >
                <ForwardRef(ButtonBase)
                  className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-disableElevation"
                  classes={
                    Object {
                      "disabled": "Mui-disabled",
                      "focusVisible": "Mui-focusVisible",
                      "root": "MuiButtonBase-root",
                    }
                  }
                  component="button"
                  disabled={false}
                  focusRipple={true}
                  focusVisibleClassName="Mui-focusVisible"
                  onClick={[Function]}
                  type="button"
                >
                  <button
                    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-disableElevation"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiButton-label"
                    >
                      Retry
                    </span>
                    <WithStyles(memo)
                      center={false}
                    >
                      <ForwardRef(TouchRipple)
                        center={false}
                        classes={
                          Object {
                            "child": "MuiTouchRipple-child",
                            "childLeaving": "MuiTouchRipple-childLeaving",
                            "childPulsate": "MuiTouchRipple-childPulsate",
                            "ripple": "MuiTouchRipple-ripple",
                            "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                            "rippleVisible": "MuiTouchRipple-rippleVisible",
                            "root": "MuiTouchRipple-root",
                          }
                        }
                      >
                        <span
                          className="MuiTouchRipple-root"
                        >
                          <TransitionGroup
                            childFactory={[Function]}
                            component={null}
                            exit={true}
                          />
                        </span>
                      </ForwardRef(TouchRipple)>
                    </WithStyles(memo)>
                  </button>
                </ForwardRef(ButtonBase)>
              </WithStyles(ForwardRef(ButtonBase))>
            </ForwardRef(Button)>
          </WithStyles(ForwardRef(Button))>
        </div>
        <div
          className="overlay"
          onClick={[Function]}
        />
      </div>
    </ScheduleLogs>
  </Connect(ScheduleLogs)>
</Provider>
`;

exports[`<ScheduleLogs /> render correctly when schedule logs has an array data and data was saved sucessfully 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ScheduleLogs)>
    <ScheduleLogs
      fetchScheduleLogs={[Function]}
      isSelected={false}
      logs={false}
      resetSelectedSchedule={[Function]}
      scheduleName={false}
      status={1}
      texts={
        Object {
          "buttonRetryText": "Retry",
          "emptyMessageText": "No schedules logs available.",
          "errorMessageText": "Something went wrong ...",
          "noSelectedScheduleText": "Please select a schedule.",
        }
      }
    >
      <div
        className="info"
      >
        <p>
          Please select a schedule.
        </p>
      </div>
    </ScheduleLogs>
  </Connect(ScheduleLogs)>
</Provider>
`;

exports[`<ScheduleLogs /> render correctly when schedule logs is empty and data was saved sucessfully 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ScheduleLogs)>
    <ScheduleLogs
      fetchScheduleLogs={[Function]}
      isSelected={false}
      logs={false}
      resetSelectedSchedule={[Function]}
      scheduleName={false}
      status={1}
      texts={
        Object {
          "buttonRetryText": "Retry",
          "emptyMessageText": "No schedules logs available.",
          "errorMessageText": "Something went wrong ...",
          "noSelectedScheduleText": "Please select a schedule.",
        }
      }
    >
      <div
        className="info"
      >
        <p>
          Please select a schedule.
        </p>
      </div>
    </ScheduleLogs>
  </Connect(ScheduleLogs)>
</Provider>
`;

exports[`<ScheduleLogs /> render correctly when schedule logs is undefined and data was saved sucessfully 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ScheduleLogs)>
    <ScheduleLogs
      fetchScheduleLogs={[Function]}
      isSelected={false}
      logs={false}
      resetSelectedSchedule={[Function]}
      scheduleName={false}
      status={1}
      texts={
        Object {
          "buttonRetryText": "Retry",
          "emptyMessageText": "No schedules logs available.",
          "errorMessageText": "Something went wrong ...",
          "noSelectedScheduleText": "Please select a schedule.",
        }
      }
    >
      <div
        className="info"
      >
        <p>
          Please select a schedule.
        </p>
      </div>
    </ScheduleLogs>
  </Connect(ScheduleLogs)>
</Provider>
`;
